### ISSUE TITLE
Structured documentation for Neural-LAM

### ISSUE THREAD

**Issue by joeloskarsson** [mllam/neural-lam#61]
As Neural-LAM grows and we make changes to it there is an increasing need for documentation. Already now the README is quite long, and some things are not that easy to find in it. I propose that we should set up a more robust and structured documentation solution. The idea of this issue is to start a discussion about how to do this.

When thinking about documentation for different open source python project, one that I really like is the [pytorch geometric documentation](https://pytorch-geometric.readthedocs.io/en/latest/). This uses [Read the docs](https://about.readthedocs.com/?ref=readthedocs.com) + [Sphinx](https://www.sphinx-doc.org/en/master/). In particular, you can create a documentation that contains both:

1. Manually written tutorials, quickstart-guides, installation instructions, etc.
2. A full API reference for all classes and functions in the codebase (descriptions + arguments + what is returned), generated directly from the docstrings in the python code.

I think having both of these is desirable, and think we could set up something like this for Neural-LAM.

I know that in [weather-model-graphs](https://github.com/mllam/weather-model-graphs) you have these nice Jupyter books for documentation @leifdenby. Is there a way to do these things in those as well? In particular, is it possible to do the automatic reference generation from docstrings?

**Comment by leifdenby** [comment]
> When thinking about documentation for different open source python project, one that I really like is the [pytorch geometric documentation](https://pytorch-geometric.readthedocs.io/en/latest/). This uses [Read the docs](https://about.readthedocs.com/?ref=readthedocs.com) + [Sphinx](https://www.sphinx-doc.org/en/master/). In particular, you can create a documentation that contains both:
> 
>     1. Manually written tutorials, quickstart-guides, installation instructions, etc.
> 
>     2. A full API reference for all classes and functions in the codebase (descriptions + arguments + what is returned), generated directly from the docstrings in the python code.

Agreed! This sounds great @joeloskarsson. I had in particular thought that around preparing datasets might be a nice thing to detail. And discuss the code structure overall too so people know where to go and change things.

> I know that in [weather-model-graphs](https://github.com/mllam/weather-model-graphs) you have these nice Jupyter books for documentation @leifdenby. Is there a way to do these things in those as well? In particular, is it possible to do the automatic reference generation from docstrings?

Thanks :) Jupyter books actually build on top of sphinx, so it might be possible to combine the use of notebooks with automatically generated docs from static code analysis with sphinx. I can look into that if that would be helpful?

**Comment by joeloskarsson** [comment]
Feel free to look into it @leifdenby, and post your findings here. I plan to also dig deeper into this, to learn a bit more about how the reference generation from docstrings actually works.

**Comment by varma1221** [comment]
Hi! I've been exploring this issue and the repository. I noticed `docs/notebooks` folder currently contains a notebook but no documentation system is set up yet.
I've been looking into Sphinx + Read the Docs as the documentation steup. Since the codebase already has well-structured docstrings, generating an API reference with `autodoc` should be fairly straightforward. I've also looked into integrating Jupyter Book for tutorials since Jupyter Book builds on Sphinx, this could naturally include the Hello World notebook discussed in #69. 
I'd be happy to put together a small proof-of-concept PR to demonstrate a possible structure so we can iterate from there. 
Let me know if this direction aligns with your expectations. Could I be assigned to this issue.

**Comment by amanmurari** [comment]
Hi, I would like to work on this issue as part of preparing for GSoC.
Could I be assigned to this?

**Comment by nikas-belogolov** [comment]
Hi, I would like to be assigned to this, sent a message about this yesterday on slack

**Comment by PES2UG23CS205** [comment]
Hi! I’ve opened a small proof-of-concept PR to explore a Sphinx documentation setup and autodoc support. I’d love to get feedback on whether this direction aligns with what you had in mind for this issue.

**Comment by PES2UG23CS205** [comment]
Hi! I created a small Sphinx proof-of-concept PR while exploring this issue and wanted to share the motivation behind the approach.

My main goals with this PoC were:

• Provide a minimal Sphinx setup that can serve as a starting point for a structured documentation system  
• Enable automatic API reference generation using existing Python docstrings via `autodoc`  
• Allow documentation to build without requiring the full ML training environment by mocking heavy dependencies  
• Keep the setup minimal so it can be iterated on together (theme, tutorials, CI integration, Jupyter Book integration, etc.)

I chose Sphinx because it is widely used for Python scientific projects and already supports:
– automatic API documentation generation  
– integration with ReadTheDocs  
– compatibility with Jupyter Book (since Jupyter Book builds on top of Sphinx)

My intention with this PR is not to finalize the documentation structure, but to provide a starting point that we can refine together.

I’d really appreciate feedback on whether this direction aligns with what you had in mind 🙂

**Comment by abhijith87x** [comment]
Hi can i work on this issue? thankyou

**Comment by Jayant-kernel** [comment]
@leifdenby @joeloskarsson 
I reviewed the recent discussion and the linked PRs (https://github.com/mllam/neural-lam/pull/196 and https://github.com/mllam/neural-lam/pull/230).

It looks like both PRs are already covering the Sphinx/docs scaffold direction, so I don’t want to duplicate that work. The main gaps I still see are:

Rendered docs preview (so maintainers can review the generated output easily)
Docs CI integration (build docs automatically in PRs)
PR preview workflow / temporary docs URL (if supported, as mentioned in the PR discussion)
I’d be happy to work on a small follow-up PR focused on this docs pipeline part.

**Comment by Mohit-Lakra** [comment]
Working on the documentation prototype for this issue.

Opened PR #252  with initial docstring improvements across core modules
these are the functions that show up most in the sphinx-autoapi output with 
incomplete documentation. Coverage: 67.5% → 69.2%.

**Comment by PES2UG23CS205** [comment]
Hi! Just wanted to follow up and see if you had any feedback on the documentation PoC draft PR. 

I’d be happy to continue iterating on the approach and pick up additional documentation-related tasks while we discuss the direction

**Comment by Mohit-Lakra** [comment]
@joeloskarsson  @leifdenby — submitted PR #272 with a Jupyter Book + sphinx-autoapi prototype.
Live demo: https://mohit-lakra.github.io/neural-lam/autoapi/index.html
The API reference is fully auto-generated from existing docstrings in neural_lam/ — no manually maintained .rst files. This directly answers the "is it possible to do automatic reference generation in Jupyter Book?" question from the original discussion.
Also have PR #252 open with docstring improvements (67.5% → 69.2% coverage) to improve the auto-generated output.

**Comment by Omnagar** [comment]
Hi 👋

I’ve been exploring the repository and the idea of automatic documentation generation. I set up a small local prototype using Sphinx with autodoc + napoleon and was able to successfully generate API documentation for several core modules such as neural_lam.utils, train_model, create_graph, and weather_dataset.

The existing docstrings seem structured enough for autodoc to extract function signatures, return types, and examples effectively. This suggests that introducing a Sphinx-based documentation system could be a viable direction.

Before proceeding further, I would like to understand:

Is Sphinx (or another tool like MkDocs) the preferred direction for the project?

How do you envision integrating documentation generation into the CI/CD workflow (e.g., GitHub Actions + Pages deployment)?

Would you prefer focusing first on API auto-generation, or also restructuring higher-level developer guides and tutorials?

I’d be happy to continue experimenting and contribute a small prototype PR if this direction aligns with your expectations.

Looking forward to your thoughts!

**Comment by sidhantpande** [comment]
Hi all 👋  

I’ve been following the discussion around structured documentation in this issue and the related PRs (#230, #252, #272, #286). Before commenting, I went through the current repository structure, existing docstrings, and the various documentation prototypes proposed so far (Sphinx scaffold, Jupyter Book + sphinx-autoapi, RTD integration, CI workflows, and docstring coverage enforcement).

From what I understand, there seems to be convergence toward:

- **Jupyter Book** as the high-level documentation framework  
- **sphinx-autoapi** for fully automated API reference generation  
- **NumPy-style docstrings** with 100% coverage enforced via pre-commit  
- **CI integration** for preview builds and deployment  

This addresses the **infrastructure layer** of documentation very well — especially the automatic API reference generation that motivated this issue.

---

## Observation

After reviewing the current codebase structure (e.g., `datastore`, `weather_dataset`, `interaction_net`, `models`, `metrics`, and the training pipeline), it seems that while we are building a strong auto-generated reference layer, there is still no structured **architecture-level narrative documentation** explaining:

- The high-level data flow  
  *(Datastore → Dataset → Model → Training → Metrics)*  
- The abstraction boundaries between modules  
- Extension points for contributors (e.g., adding new datastores, models, metrics)  
- The conceptual role of limited area modeling within the framework  
- Where a new contributor should start when modifying core components  

In other words, we are building an excellent **reference manual**, but we don’t yet have a clear **mental model guide** for developers.

The original issue description mentioned discussing the overall code structure so contributors know where to go and change things — and I believe this architecture and contributor-oriented layer might still be missing.

---

## Proposal (for alignment)

Before starting any work, I wanted to ask:

Would it be valuable to add a structured **“Architecture & Developer Guide”** section on top of the Jupyter Book setup (once finalized), focusing on:

- System overview diagrams  
- Module interaction explanations  
- Key design decisions  
- Extension patterns  

This would complement the API auto-generation work rather than overlap with it.

If this aligns with the intended direction for #61, I’d be happy to draft a small proposal outlining scope and structure before opening a PR.

Looking forward to your thoughts 🙂

**Comment by A101ROHIT** [comment]
Hi mentors @observingClouds @leifdenby @joeloskarsson,

I really like the direction suggested by @joeloskarsson about adopting a documentation system similar to PyTorch Geometric using **Sphinx + ReadTheDocs**, combining structured guides with automatically generated API documentation.

While exploring the Neural-LAM repository and reading through the README, I noticed that it already contains a lot of useful information. However, as the project grows, the README may become harder to navigate and maintaining it manually could require significant effort. A structured and automated documentation system could help keep information organized while ensuring the documentation stays synchronized with the codebase.

### Possible Direction

_**1. Automatic API documentation from `neural_lam`**_

Since many parts of the codebase already contain docstrings, tools such as **Sphinx (with autodoc/AutoAPI)** could generate a full API reference directly from the source code.

This would automatically create documentation pages for modules like:

`neural_lam.data`
`neural_lam.models`
`neural_lam.training`
`neural_lam.graph`

The benefit is that the API documentation would remain **automatically synchronized with the code**, reducing manual maintenance.

_**2. Combining notebooks with the documentation system**_

Since **Jupyter Book builds on top of Sphinx**, it might integrate well with the current repository. This could allow us to include the notebooks present in `docs/notebooks` as interactive tutorials while still maintaining structured API documentation.

_**3. Fully automated documentation pipeline (important)**_

It might also be valuable to make documentation generation **fully automated** through `CI`. For example:

code change → CI builds documentation → API reference updates automatically

This would ensure that the documentation evolves alongside the codebase without requiring manual updates.

### Next Step

As a next step, I want to explore other options than Sphinx and  along with buiding a prototype that:

* sets up a **Sphinx-based documentation structure**
* automatically generates API documentation from the `neural_lam` package
* explores integrating notebooks/tutorials into the documentation and CI pipeline.

I also want to explore the point mentioned by @leifdenby about providing clearer documentation around **dataset preparation and overall code structure**, which could be very helpful for new users.

If this direction sounds useful, I’d be happy to experiment with a prototype and share the results here for feedback.

**Comment by SaiKushwant** [comment]
Hi  Neural-LAM team,

I’ve been following the progress on the documentation prototype and PR #252. I am very interested in contributing to Neural-LAM for GSoC 2026.

I’ve successfully set up the environment on Ubuntu using uv and have been exploring the BaseDataStore and WeatherDataset abstractions. I’m particularly impressed by how the repository modularizes the graph structures (like the hierarchical mesh for Hi-LAM) from the training logic.

Since the current docstring coverage is at 69.2%, I’d like to help bridge that gap—specifically by documenting the m2m_edge_index and mesh_up/down logic in the hierarchical models, as these are crucial for new contributors to understand the message-passing flow.

Is there a specific "Good First Issue" or a module with low coverage you'd like me to prioritize to demonstrate my fit for the project?

**Comment by joeloskarsson** [comment]
@SaiKushwant maybe you can go over #252 and see if there are any docstrings there that can be improved? That would be very helpful for the community. My understanding is that that PR will take us to 100% docstring coverage at least, but we need to review that all the comments are accurate.

**Comment by aishwaryabnateekar-ui** [comment]
Hi! I’m Aishwarya, a beginner contributor interested in improving documentation. 
I’m currently exploring the repository and would love to help with this issue if possible. 
Please let me know if I can contribute or start with a small part of the documentation.

**Comment by sadamov** [comment]
@aishwaryabnateekar-ui hi! as joel already mentioned above, it's best to head over to PR #252 and improve the existing PRs there.

**Comment by koffandaff** [comment]
Hi @joeloskarsson  @leifdenby  following up on my introduction done in #gsoc-project2(slack channel) — I've been working on this idea (**Automation of docs**) for about a week and wanted to share my early progress.

I've been building AutoCodeDoc [(repo),](https://github.com/koffandaff/AutoCodeDoc.git) a documentation engine designed around the kind of automated pipeline the GSOC project calls for ( Live docs at:  [Docs](https://koffandaff.github.io/AutoCodeDoc/)  )

It's still actively under development and testing, but it already includes automated API reference generation from docstrings, architecture and ER diagrams generated via 

- AST parsing of Python imports and Pydantic models
-  a documentation integrity checker that blocks builds when fields are undocumented
-  a single-command pipeline (run_pipeline.py),
- a GitHub Actions CI/CD workflow that deploys to GitHub Pages on every push to main.


The stack is Python, FastAPI, Pydantic v2, MkDocs with the Material theme, mkdocstrings, and Mermaid.js for diagrams.

I'd love to hear your thoughts:
- does this direction align with what you have in mind for the neural-lam documentation system? 
-  Are there specific aspects of the codebase or workflow I should be prioritising? 
- Any features that should be present or any noise that should be cancelled out

Any feedback on the architecture or approach would be very welcome.

willing to get assigned and implement this in Neural-LAM 

> As Neural-LAM grows and we make changes to it there is an increasing need for documentation. Already now the README is quite long, and some things are not that easy to find in it. I propose that we should set up a more robust and structured documentation solution. The idea of this issue is to start a discussion about how to do this.
> 
> When thinking about documentation for different open source python project, one that I really like is the [pytorch geometric documentation](https://pytorch-geometric.readthedocs.io/en/latest/). This uses [Read the docs](https://about.readthedocs.com/?ref=readthedocs.com) + [Sphinx](https://www.sphinx-doc.org/en/master/). In particular, you can create a documentation that contains both:
> 
> 1. Manually written tutorials, quickstart-guides, installation instructions, etc.
> 2. A full API reference for all classes and functions in the codebase (descriptions + arguments + what is returned), generated directly from the docstrings in the python code.
> 
> I think having both of these is desirable, and think we could set up something like this for Neural-LAM.
> 
> I know that in [weather-model-graphs](https://github.com/mllam/weather-model-graphs) you have these nice Jupyter books for documentation [@leifdenby](https://github.com/leifdenby). Is there a way to do these things in those as well? In particular, is it possible to do the automatic reference generation from docstrings?

**Comment by AsamaeS** [comment]
Hi! I'm Asmae, an Innovation & Amoa engineering student preparing a GSoC proposal for Neural-LAM.

I’ve been reading through the discussion here and exploring the repository structure. The idea of combining a structured documentation system with automatically generated API references from docstrings looks very promising.

As a first step, I’m planning to experiment locally with a small Sphinx setup to see how much of the current neural_lam package can be automatically documented from existing docstrings.

If that sounds useful, I’d be happy to share a small prototype or observations once I have something working.

**Comment by Gauri2425** [comment]
Hello everyone,

I’m a beginner contributor who is very interested in working with the **Neural-LAM** project and learning from the community. I recently went through the discussion about improving the documentation, and I really liked the idea of creating a more structured documentation system instead of relying mainly on a long README.

From what I understand, as the project grows it becomes harder for new contributors and users to quickly find things like installation steps, tutorials, or explanations of different classes and functions. Having a proper documentation structure could make it much easier for people to explore and contribute to the project.

One approach that looks promising to me is using **Sphinx** together with **Read the Docs**, similar to how the documentation is organized in **PyTorch Geometric**. My understanding is that this would allow us to include two important parts in the documentation:

1. User-focused guides such as installation instructions, quickstart tutorials, and examples of training or running models.
2. Automatically generated API documentation created from Python docstrings, so that descriptions of classes and functions stay synchronized with the code.

I am still learning and exploring the codebase, but I would love to try setting up a small prototype documentation structure to better understand how this could work for the project. For example, I could experiment with generating API documentation from docstrings and organizing a few basic guide pages.

I would really appreciate any suggestions on where a beginner like me could start contributing in this area. I’m excited to learn more about the project and gradually become more involved with the community.

Thank you!

**Comment by sadamov** [comment]
Hi @AsamaeS, hi @Gauri2425 and welcome! There are already some PRs open that tackle this issue. 
To improve the docstrings: #252 
And as you plan to implement sphinx+readthedocs: #272 (closed: #196 #230)
So instead of starting from scratch feel free to go over to either of these two PRs and review/discuss the solutions there.

**Comment by Seai5** [comment]
Hello! I’m Hrithik, excited to contribute and already active on the Slack channels

Here’s what I’ve done so far:

Set up the repo with uv and an editable install.

Planning to run a 1-epoch training today.

I noticed that recent PRs (#428 Sphinx POC and #322 Jupyter Book) were closed due to size or duplication. I understand the focus is on improving existing work, and I’d love to contribute in that spirit.

Specifically, I would like to:

Test #428 locally.

Clean up 1–2 docstrings.

Make a small, clean Sphinx + autodoc PR.

Would it be possible for me to be assigned to #61? I’ll share my local POC in Slack tomorrow for feedback.

Looking forward to contributing!

Thank you

**Comment by sumit08kumar** [comment]
Hi, I have submitted a PR with an initial prototype for structured documentation using Sphinx.

Would appreciate feedback on the approach and structure.
PR link: https://github.com/mllam/neural-lam/pull/501

Thank you!
@leifdenby @joeloskarsson

**Comment by Chaitanya-Parab** [comment]
Hi @joeloskarsson ,

I’ve been going through this discussion and I really like the idea of moving toward a more structured documentation system as Neural-LAM grows.

A simple and practical approach we could take is to start with **Sphinx + Read the Docs**, since it naturally supports both:
 1. Auto-generating API references from existing Python docstrings.
 2. And writing structured guides like tutorials, installation and quickstart docs.

My proposed approach would be incremental (to keep it manageable):
1. Start small (foundation)
     * Set up a basic Sphinx configuration.
     * Enable `autodoc` + `napoleon` to generate API docs from current docstrings.
     * This gives us immediate value without large refactoring.

2. Improve structure gradually
    * Break down the current README into sections.
    * Move them into proper documentation pages with better navigation.

3. Add automation
    * Integrate docs build into CI.
    * Optionally deploy via Read the Docs so documentation stays updated automatically.

Regarding Jupyter boolk I think it’s a great option for tutorials and examples. If needed, we could explore integrating notebooks into Sphinx using tools like MyST, so we get the best of both worlds without overcomplicating the setup early on. My goal would be to keep the solution simple, maintainable and easy for contributors to extend, rather than introducing too many tools at once.

Would love to hear your thoughts.

Thank You
Chaitanya Parab

**Comment by codedbyasim** [comment]
Hi @joeloskarsson , I would love to help out with this documentation effort as part of my GSoC preparation!
I noticed that for autodoc or napoleon to successfully generate the automatic API references, several core modules are completely missing proper NumPy-style parameter docstrings.
I have already gone ahead and formatted/added these missing NumPy-style docstrings locally for:

1. The core graph models (BaseGraphModel, GraphLAM, and HiLAM)
2. The standard dataset module (WeatherDataModule in weather_dataset.py)

Could I go ahead and open PR(s) for these docstring additions? This will ensure the automatic API reference pages populate correctly once the main Sphinx configuration is merged.

**Comment by NikhilTiwari-hub** [comment]
Hi, I am interested in working on improving the documentation system for Neural-LAM.
I have explored tools like Sphinx and ReadTheDocs, and I would like to contribute by setting up an initial documentation structure and auto-generating API references from docstrings.
Could you please guide me on how to get started or if there are any specific parts I should focus on first?

**Comment by sanketpatilm5** [comment]
Hi Joel, Leif — I've been looking into this and wanted to share where I've landed technically.

Jupyter Book is fully compatible with autodoc and napoleon since it compiles down to Sphinx internally. I've verified this locally on a fork — every Sphinx extension works natively, including sphinx.ext.viewcode for source links and sphinx.ext.intersphinx for cross-references to PyTorch/xarray docs.

The main CI challenge is neural-lam's heavy dependency chain (torch, torch_geometric, pytorch_lightning, cartopy, etc.). Using autodoc_mock_imports solves this cleanly — mock imports create stub objects so all modules import successfully and docstrings are extracted without needing a GPU stack. I got a full build passing on a standard environment in under 2 minutes.

For the API surface, I've mapped out what needs autodoc coverage:

neural_lam.datastore — BaseDatastore abstract methods, BaseRegularGridDatastore, MDPDatastore, NpyFilesDatastoreMEPS
neural_lam.models — the BaseGraphModel → GraphLAM and BaseHiGraphModel → HiLAM → HiLAMParallel hierarchy
neural_lam.config — NeuralLAMConfig fields, load_config_and_datastore()
neural_lam.weather_dataset, neural_lam.create_graph, and remaining modules
For CI, a .github/workflows/docs.yml running jupyter-book build --all docs/ --nitpick on PRs (promoting warnings to errors) would catch broken references before merging, alongside the existing install-and-test.yml.

I've run interrogate against the current codebase to get a docstring coverage baseline — happy to share the numbers if useful. One question: is there a preference between GitHub Pages and Read the Docs for hosting?

I'm working on a GSoC 2026 proposal around this — would appreciate any feedback on the approach.

**Comment by Mheet** [comment]
I've been following this thread and wanted to add something that hasn't really come up yet, Docstring coverage enforcement in CI.
Most of the PRs here focus on the docs tooling itself, which makes sense. But without a CI gate, coverage will just drift again as new code gets added. What I'm thinking is a lightweight ast-based checker that walks the source tree, flags any public function or class missing a docstring, and fails the PR. No external dependencies, just stdlib.
This would complement whatever tooling direction gets chosen (#272, #456, etc.) since the enforcement layer is pretty much orthogonal to the rendering stack. Happy to draft a proof-of-concept if that seems useful.

**Comment by 2024itb047samata** [comment]
Hi! I’m interested in working on structured documentation for Neural-LAM as part of GSoC.

I’ve gone through the issue and I’m planning to explore tools like Sphinx/ReadTheDocs to generate API documentation from docstrings and create a structured documentation system.

I would love to get your guidance and feedback on the approach. Also happy to start with smaller contributions.

Thanks!

**Comment by ManviRankawat** [comment]
Hi @joeloskarsson, I’ve submitted my GSoC proposal for automating API documentation for Neural-LAM. In preparing it, I explored the weather-model-graphs code and design, especially the use of networkx.DiGraph for modular graph construction and splitting via edge attributes like g2m, m2m, and m2g.

This helped me frame the proposal around generating structured API docs from Python docstrings while keeping high-level guides for tutorials and installation, reflecting the code structure and graph components clearly.

Looking forward to your feedback!
